testing.md

```diff
@@ -15,10 +15,56 @@ as well as always iterating on our testing infrastructure.
 
 ## Testing Framework
 
-[`py.test`](https://docs.pytest.org/en/latest/) is the default framework we use
+[`pytest`](https://docs.pytest.org/en/latest/) is the default framework we use
 for testing.
 
 
+## Test Suite Structure
+
+### Test Module Naming/Organization
+
+Where possible, test suites should mirror the module structure of their
+corresponding package.  For example, the
+[`eth-abi`](https://github.com/ethereum/eth-abi) package has a submodule called
+[`eth_abi.grammar`](https://github.com/ethereum/eth-abi/blob/master/eth_abi/grammar.py).
+Accordingly, it has a testing module called
+[`tests.grammar`](https://github.com/ethereum/eth-abi/blob/master/tests/grammar.py)
```

Comment:
maybe don't use dotted module representation here, but instead file paths like `./tests/test_grammar.py`

Reply:
I think this is directing that any tests on the top level `eth_abi` go straight in the `tests/` folder.

One issue with that is packages that have multiple top-level imports (like `ens` in web3.py). It's solved in there by using `tests/core` (for web3 things), `tests/ens` for ens things and `tests/integration`. We have e2e and unit tests co-mingled at the moment. So we could add a fourth `tests/end_to_end` which is presumed to be for web3. I guess we would add a `tests/ens_end_to_end` also. We shouldn't see a huge prolifiration of secondary modules in packages, so this may not be crazy...

Reply:
Yeah, actually I was thinking that maybe it's better to have the top-level test module structure look something like what you're saying.  So maybe `tests.core` is where the tests go that would mirror the package structure.  Those would be unit tests or just more direct tests.  Then tests that compose lots of different parts could go in `tests.end_to_end`.

Reply:
But then there's the problem that there's often a top-level submodule called `core` in many packages.

Reply:
Do we actually have that in many of ours.  My memory says not.

Reply:
@pipermerriam Part of my reasoning for using dot paths is that it's more correct to say that the "name" of a module is its dot path rather than its file path.

Reply:
@pipermerriam Yeah, I guess not actually.  For some reason I thought we did.